### ISSUE TITLE
Improve performance of Month class>>indexOfMonth:

### ISSUE THREAD

**Issue by chisandrei** [guillep/pharo-core#10676]
When parsing the `DateAndTime` objects the method `Month class>>indexOfMonth:` is called. This ends up doing:

```
Month class>>#indexOfMonth: aMonthName 
	1 to: 12 do: [ :i |  (aMonthName, '*' match: (MonthNames at: i)) ifTrue: [^i] ].
 	self error: aMonthName , ' is not a recognized month name'.
```

`aMonthName` however is a symbol as defined in `ChronologyConstants`. This means `aMonthName, '*'` creates a `Symbol` which is slow.

A different way to implement this method would be as below. This does not use a symbol conversion nor calls match:

```
Month class>>#indexOfMonth: aMonthName 
	1 to: 12 do: [ :i | ((MonthNames at: i) beginsWith: aMonthName) ifTrue: [^i] ].
 	self error: aMonthName , ' is not a recognized month name'.
```

More details at https://github.com/feenkcom/gtoolkit/issues/2310

**Comment by guillep** [comment]
Thanks for catching that!

I have two comments:
 - first, using `beginsWith:` changes a bit the semantics: `match:` is case insensitive, while the `beginsWith:` is case sensitive. See for example: `'asd' match: 'ASD' "true"`
 - the other, I think your proposal can be further enhanced by just using an `indexOf:`. In my machine doing something like `1 to: 12 do: [ :i | ((MonthNames indexOf: aMonthName) ~= 0) ifTrue: [^i] ].` is in average ~4x faster than your solution, but still losing the case insensitivity

Now, as it is this is a an API breaking change, in the core image there are not many users, but It's difficult to say if there are external users, so to take with care.

**Comment by chisandrei** [comment]
Indeed, `match:` is cases insensitive. Made another change where I switched to `beginsWith:caseSensitive:`.  

But also this still introduces an API change as something like `Month indexOfMonth: '*ebruar'` will no longer work. Not sure if there are any users of that, or if that behaviour is wanted.

The smallest change that would avoid creating symbols would be `1 to: 12 do: [ :i |  (aMonthName asString, '*' match: (MonthNames at: i)) ifTrue: [^i] ].`; so just calling `asString` to avoid the symbol creation. But also using `beginsWith:caseSensitive:` seems reasonable.

**Comment by chisandrei** [comment]
Interestingly the current slowdown was when reading `DateAndTime` objects, like `DateAndTime fromString: DateAndTime now asString`. This ends up here: https://github.com/pharo-project/pharo/blob/37934d7c40480f096548689ab93d4cad88f0344b/src/Kernel-Chronology-Extras/Date.extension.st#L270 

So seems the parser has the index of the month, gets the name of the month and then later on gets the index again.
Both `Date year: 2018 month: 9 day: 15` and `Date year: 2018 month: #September day: 15` work, but when passing the name there is an extra call to get the index.

Apart from this, or even instead, `Date class>>#readFrom:` could be updated to not get the name of the month when it already has the index.

**Comment by MarcusDenker** [comment]
Has been merged!

What would help: Some people doing PRs are actually working on projects using Pharo as part of a team. 

If one person of the team does a PR, it is *very much a good idea* if another person of the team reviews that PR.

Eg here @chisandrei did the PR, but @syrel looked at it and decided it is good enough to use internally. To me, that is exactly what a "review" is about: "Ah yes, I looked at it and I think it is good."

 If @syrel would have reviewed the PR, it would have been merged months ago. Without it, the PR waits for someone else to look at it. 

I try my best, but my time is quite limited that I can spend on it.
